### PR TITLE
Adding charge calculation to Volta mini app [volta-charge-dev]

### PR DESCRIPTION
**Pull request by mlstowell**
This is primarily a handful of changes to the volta miniapp to compute total charge from the dielectric flux.  This is done in two different ways as a demonstration of performing volume and surface integrals of L2 and RT fields respectively.  The two results are identical to machine precision but users may prefer one method over the other for reasons of efficiency etc..  

The surface integral required adding a new `LinearFormIntegrator` called `VectorFEBoundaryLFIntegrator`.  This integrator does not take any arguments in its constructor.  It should only be used with H(Div) bases and it simply integrates the normal component of the basis vector over each boundary face.  The resulting linear form, when acting on an H(Div) GridFunction, computes the surface integral of the normal component of the H(Div) field over the given boundary.

| PR | Author | Editor | Reviewers  | Assignment | Approval | Merge |
| --- | --- | --- | ---  | --- | --- | --- |
| https://github.com/mfem/mfem/pull/994 | @mlstowell  | @tzanio | @psocratis  + @dylan-copeland | 11/23/19 | 12/12/19 | ⌛due 12/19/19 |

### CODE REVIEW

File: fem/lininteg.cpp [hunk 1]
```diff
@@ -351,6 +351,36 @@ void VectorFEDomainLFIntegrator::AssembleDeltaElementVect(
...
+   {
+      int intorder = 2*el.GetOrder();  // <----------
+      ir = &IntRules.Get(el.GetGeomType(), intorder);
+   }
+
+   for (int i = 0; i < ir->GetNPoints(); i++)
+   {
+      const IntegrationPoint &ip = ir->IntPoint(i);
+
+      Tr.SetIntPoint (&ip);
```
Comment (v-dobrev) [inline fem/lininteg.cpp]: `Tr` is not used, so this line can be removed.

File: fem/lininteg.cpp [hunk 2]
```diff
@@ -351,6 +351,36 @@ void VectorFEDomainLFIntegrator::AssembleDeltaElementVect(
...
+      double val = ip.weight;
+
+      el.CalcShape(ip, shape);
+
+      add(elvect, val, shape, elvect);
```
Comment (v-dobrev) [inline fem/lininteg.cpp]: This can be shortened to `elvect.Add(val, shape)`.

File: fem/lininteg.cpp [hunk 3]
```diff
@@ -351,6 +351,36 @@ void VectorFEDomainLFIntegrator::AssembleDeltaElementVect(
...
+   int dof = el.GetDof();
+
+   shape.SetSize(dof);
+   elvect.SetSize(dof);
+   elvect = 0.0;
+
+   const IntegrationRule *ir = IntRule;
+   if (ir == NULL)
+   {
+      int intorder = 2*el.GetOrder();  // <----------
```
Comment (v-dobrev) [inline fem/lininteg.cpp]: For this integrator, it should be sufficient to use `intorder = el.GetOrder()` and even `intorder = el.GetOrder()-1` for the standard RT spaces.

**Comment by mlstowell** [comment]
All excellent points.  Thanks, @v-dobrev !  I'll make those changes and double-check the results.

**Comment by mlstowell** [comment]
@dylan-copeland pointed out that the `VectorFEBoundaryLFIntegrator` is just a special case of the `VectorFEBoundaryFluxLFIntegrator` and the differences were not very significant.  Consequently, I have removed this new integrator in favor of generalizing the existing integrator by making the `Coefficient` optional.  This is more in keeping with the `BilinearFormIntegrators` which often accept optional coefficient arguments.

**Comment by tzanio** [comment]
> Consequently, I have removed this new integrator in favor of generalizing the existing integrator by making the Coefficient optional.

Should https://mfem.org/lininteg/#vector-field-operators be updated to reflect this?

**Comment by mlstowell** [comment]
Thanks, @dylan-copeland! I missed that factor of 2.

**Comment by tzanio** [comment]
Reminder: the reviews for this PR are due this week

cc: @psocratis, @dylan-copeland

**Comment by mlstowell** [comment]
> https://mfem.org/lininteg/#vector-field-operators

No, I don't think the `VectorFEBoundaryLFIntegrator` made it into the documentation (oops).  So, no need to remove it.

**Comment by dylan-copeland** [comment]
I am approving this PR although the tests failed for me and @psocratis. There must be something incorrect in our testing procedure, as @psocratis and I both got differences in all test results, although this PR should not affect those tests. It took over a week to get the tests to run successfully, and now that the results are wrong, it seems better to rely on someone more experienced (e.g. @tzanio) to test.

**Comment by psocratis** [comment]
I agree with @dylan-copeland. Baseline tests fail by giving different number of iterations for almost all examples (including serial runs)
I suggest further investigation is needed.

**Comment by vladotomov** [comment]
Seems that master hasn't been merged in a while.

**Comment by mlstowell** [comment]
That could explain things...  I should have thought of that.  I'll take care of it.

**Comment by psocratis** [comment]
I will rerun the tests before approving.

**Comment by tzanio** [comment]
@dylan-copeland do you approve this?

**Comment by dylan-copeland** [comment]
The tests finally ran successfully, except for one issue with ex13p (see below). I manually ran all the sample runs for ex13p and did not see any failures like this, so I don't know how this failure occurred. Assuming for now that this is an issue with testing and not the PR, I will approve the PR.

@@ -80942,10 +80942,10 @@ <examples> mpirun -np 4 ex13p -m ../data
 Iteration 30 	bsize 5 	maxres   5.32436016525620e-04
 Iteration 31 	bsize 5 	maxres   5.25630381369916e+00
 Iteration 32 	bsize 5 	maxres   1.17068386877493e+06
-DPOTRF INFO = 4
-
 Error in LOBPCG:
 Orthonormalization of residuals failed

**Comment by tzanio** [comment]
Merged in `next` for testing.

**Comment by v-dobrev** [comment]
The issue that @dylan-copeland saw is related to the random initial guess for LOBPCG, I suspect. The example is very sensitive to the random seed, at least with older versions of hypre; with newer versions of hypre this sensitivity is not that bad. Maybe this is the reason (using a newer version of hypre) that running the sample runs manually was not producing errors.

**Pull request merged**